govcd/api_token.go

```diff
@@ -21,122 +21,208 @@ import (
+	}
+
+	queryParameters := url.Values{}
+	queryParameters.Add("filter", fmt.Sprintf("(name==%s;owner.name==%s;(type==PROXY,type==REFRESH))", tokenName, userName))
+
+	tokens, err := vcdClient.GetAllTokens(queryParameters)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get token by name and owner: %s", err)
 	}
-	var userDef string
-	newUrl := new(url.URL)
-	newUrl.Scheme = vcdClient.Client.VCDHREF.Scheme
-	newUrl.Host = vcdClient.Client.VCDHREF.Host
-	urlStr := newUrl.String()
+
+	token, err := oneOrError("name", tokenName, tokens)
+	if err != nil {
+		return nil, err
+	}
+
+	return token, nil
+}
+
+func (vcdClient *VCDClient) RegisterToken(org string, tokenParams *types.ApiTokenParams) (*types.ApiTokenParams, error) {
+	client := vcdClient.Client
+
+	// If the client is a user of an org, the endpoint is oauth/tenant/orgName/register
+	// if the client is a user of sysorg, the endpoint is oauth/provider/register
+	userDef := "tenant/" + org
 	if strings.EqualFold(org, "system") {
 		userDef = "provider"
-	} else {
-		userDef = fmt.Sprintf("tenant/%s", org)
 	}
-	reqUrl := fmt.Sprintf("%s/oauth/%s/token", urlStr, userDef)
-	reqHref, err := url.ParseRequestURI(reqUrl)
+
+	// Create the URL for the register endpoint
+	urlRef, err := url.ParseRequestURI(fmt.Sprintf("%s://%s/oauth/%s/%s", client.VCDHREF.Scheme, client.VCDHREF.Host, userDef, "register"))
 	if err != nil {
-		return nil, fmt.Errorf("error getting request URL from %s : %s", reqUrl, err)
+		return nil, fmt.Errorf("error getting request URL from %s : %s", urlRef.String(), err)
 	}
 
-	data := bytes.NewBufferString(fmt.Sprintf("grant_type=refresh_token&refresh_token=%s", token))
-	req := vcdClient.Client.NewRequest(nil, http.MethodPost, *reqHref, data)
-	req.Header.Add("Accept", "application/*;version=36.1")
+	newTokenParams := &types.ApiTokenParams{}
+	err = client.OpenApiPostItemSync("36.1", urlRef, nil, tokenParams, newTokenParams)
+	if err != nil {
+		return nil, fmt.Errorf("error registering token: %s", err)
+	}
+
+	return newTokenParams, nil
+}
+
+// GetAccessToken gets the access token structure containing the bearer token
+func (client *Client) GetAccessToken(org, funcName string, data *bytes.Buffer) (*types.ApiTokenRefresh, error) {
+	resp, err := client.doTokenRequest(org, "token", "36.1", "application/x-www-form-urlencoded", data)
+	if resp != nil {
+		defer resp.Body.Close()
+	}
 
-	resp, err := vcdClient.Client.Http.Do(req)
 	if err != nil {
 		return nil, err
 	}
 
-	var body []byte
-	var tokenDef types.ApiTokenRefresh
-	if resp.Body != nil {
-		body, err = io.ReadAll(resp.Body)
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
 	}
 
-	// The default response data to show in the logs is a string of asterisks
 	responseData := "[" + strings.Repeat("*", 10) + "]"
 	// If users request to see sensitive data, we pass the unchanged response body
 	if util.LogPasswords {
 		responseData = string(body)
 	}
```

Comment:
There is a centralised function `hideSensitive` that captures sensitive data and filters it

Reply:
Updated `hideSensitive` to handle the new cases for the refresh token and access token from `getAccessToken`